### ISSUE TITLE
dev loss overfits after 7k steps (SQuAD)

### ISSUE THREAD

**Issue by wenwei202** [allenai/bi-att-flow#36]
**setting**:
The dev branch code
tensorflow v1.1.0
Python 3.6.1 |Anaconda 4.4.0 (64-bit)
1 TITAN X (Pascal) 
Driver Version: 375.26 

**Running**:
Run the default scripts in the [tutorial](https://github.com/allenai/bi-att-flow/tree/dev) to duplicate the result of SQuAD.

**Results**:
```
python -m basic.cli --batch_size 30
test step 20000: accuracy=0.6044, f1=0.7255, loss=3.7624 

python squad/evaluate-v1.1.py $HOME/data/squad/dev-v1.1.json out/basic/00/answer/test-020000.json
{"exact_match": 63.434247871333966, "f1": 73.9342647706341} 
```

![dev](https://user-images.githubusercontent.com/12142066/27155435-d82b855c-510d-11e7-9565-4457fe7e835f.PNG)
![train](https://user-images.githubusercontent.com/12142066/27155436-d830d1b0-510d-11e7-8a37-caafdc6e1bcd.PNG)

Seems it overfits after 7K steps.
The score at step 7K:
```
python -m basic.cli --batch_size 30
test step 7000: accuracy=0.6129, f1=0.7289, loss=3.0944

python squad/evaluate-v1.1.py $HOME/data/squad/dev-v1.1.json out/basic/00/answer/test-007000.json
{"exact_match": 64.00189214758751, "f1": 74.19603365373315}
```
An valuable mention is issue #35 . And the default learning rate is `0.001` instead of `0.5` in the paper. I suppose `0.5` is for the CNN dataset.
I also notice this is a comment in the tutorial `Dev Data (old) NOTE: These numbers are from v0.2.1`. Why would this matter?

Any insight will be appreciated.

**Comment by insikk** [comment]
Even though there seems overfit. You should pick the best model on dev set, not the point overfit starts. The cross entropy loss is not always consistent with the accuracy as you can see here.

**Comment by wenwei202** [comment]
@insikk Thanks for those insights. The best dev/acc is at 7K while the best dev/f1 is at 14K. f1 is a more reliable one?

**Comment by insikk** [comment]
I think so. When I saw some incorrect predictions, which has em value 0 (False), it seems possible to be considered as correct answer. For example, missing/containing proposition is considered as incorrect on EM score. F1 is more reliable in my opinion for information retrieval purpose. 
However, if what you want to achieve is generating human like answer, em is better than f1.

I also tried metric for answer containing sentence prediction. 

What do you think?

**Comment by wenwei202** [comment]
Seems I was enable to duplicate the results by both tensorflow 0.1.2 and tensorflow 1.1.0 (dev branch).
What I did was:
1) cleaned up all generated datasets which had been generated before, and regenerated it
2) used cuda 8.0 and cudnn 5, which had NOT been used somehow

Also refer to #35
